File: apps/android/app/src/main/java/com/smartfind/client/net/Protocol.kt
package com.smartfind.client.net

import org.json.JSONObject

/**
 * These data classes mirror packages/protocol/src/index.ts. Kotlin can't
 * import the TypeScript package directly, so this file is the Android side
 * of a manually-synchronized contract. If you change the wire format in one
 * place, change it in both - docs/protocol.md is the source of truth for
 * the JSON shape both sides must agree on.
 */

sealed class ServerToClient {
    data class Welcome(val deviceId: String, val serverTime: Long) : ServerToClient()
    data class Command(val command: IncomingCommand) : ServerToClient()
    data class DeviceList(val raw: JSONObject) : ServerToClient()
    data class StatusUpdate(val raw: JSONObject) : ServerToClient()
    data class Error(val message: String) : ServerToClient()
}

sealed class IncomingCommand(open val requestId: String) {
    data class Ring(override val requestId: String, val target: String, val durationMs: Long) :
        IncomingCommand(requestId)
    data class StopRing(override val requestId: String, val target: String) :
        IncomingCommand(requestId)
    data class Message(override val requestId: String, val target: String, val message: String) :
        IncomingCommand(requestId)
    data class Ping(override val requestId: String) : IncomingCommand(requestId)
}

enum class CommandStatus(val wire: String) {
    SENDING("sending"),
    DELIVERED("delivered"),
    EXECUTING("executing"),
    COMPLETED("completed"),
    FAILED("failed"),
    TIMED_OUT("timed_out"),
}

object Protocol {

    fun parseServerToClient(json: String): ServerToClient? {
        val obj = JSONObject(json)
        return when (obj.optString("kind")) {
            "welcome" -> ServerToClient.Welcome(
                deviceId = obj.getString("deviceId"),
                serverTime = obj.getLong("serverTime"),
            )
            "command" -> {
                val cmd = obj.getJSONObject("command")
                val requestId = cmd.getString("requestId")
                val parsed = when (cmd.optString("type")) {
                    "RING" -> IncomingCommand.Ring(
                        requestId, cmd.getString("target"), cmd.optLong("durationMs", 30_000L)
                    )
                    "STOP_RING" -> IncomingCommand.StopRing(requestId, cmd.getString("target"))
                    "MESSAGE" -> IncomingCommand.Message(
                        requestId, cmd.getString("target"), cmd.getString("message")
                    )
                    "PING" -> IncomingCommand.Ping(requestId)
                    else -> return null
                }
                ServerToClient.Command(parsed)
            }
            "device_list" -> ServerToClient.DeviceList(obj)
            "status_update" -> ServerToClient.StatusUpdate(obj)
            "error" -> ServerToClient.Error(obj.optString("message"))
            else -> null
        }
    }

    fun hello(deviceId: String, token: String): String =
        JSONObject().apply {
            put("kind", "hello")
            put("deviceId", deviceId)
            put("token", token)
            put("deviceInfo", JSONObject())
        }.toString()

    fun ack(requestId: String, status: CommandStatus, detail: String? = null): String =
        JSONObject().apply {
            put("kind", "ack")
            put("requestId", requestId)
            put("status", status.wire)
            if (detail != null) put("detail", detail)
        }.toString()

    fun heartbeat(batteryPercent: Int?, charging: Boolean?): String =
        JSONObject().apply {
            put("kind", "heartbeat")
            if (batteryPercent != null) put("battery", batteryPercent)
            if (charging != null) put("charging", charging)
        }.toString()
}

File: apps/ios/SmartFind/Networking/Protocol.swift
import Foundation

/// Mirrors packages/protocol/src/index.ts. Swift can't import the shared
/// TypeScript package, so - same as the Android client - this file is the
/// iOS side of a manually-synchronized contract. See docs/protocol.md for
/// the canonical JSON wire shape both sides must agree on.

enum CommandStatus: String, Codable {
    case sending, delivered, executing, completed, failed
    case timedOut = "timed_out"
}

enum IncomingCommand {
    case ring(requestId: String, target: String, durationMs: Double)
    case stopRing(requestId: String, target: String)
    case message(requestId: String, target: String, text: String)
    case ping(requestId: String)
}

enum ServerToClient {
    case welcome(deviceId: String, serverTime: Double)
    case command(IncomingCommand)
    case deviceList(Data)
    case statusUpdate(Data)
    case error(String)
}

enum ProtocolCodec {

    static func parse(_ jsonData: Data) -> ServerToClient? {
        guard
            let obj = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
            let kind = obj["kind"] as? String
        else { return nil }

        switch kind {
        case "welcome":
            guard let deviceId = obj["deviceId"] as? String,
                  let serverTime = obj["serverTime"] as? Double else { return nil }
            return .welcome(deviceId: deviceId, serverTime: serverTime)

        case "command":
            guard let cmd = obj["command"] as? [String: Any],
                  let requestId = cmd["requestId"] as? String,
                  let type = cmd["type"] as? String else { return nil }
            switch type {
            case "RING":
                let target = cmd["target"] as? String ?? ""
                let duration = cmd["durationMs"] as? Double ?? 30_000
                return .command(.ring(requestId: requestId, target: target, durationMs: duration))
            case "STOP_RING":
                let target = cmd["target"] as? String ?? ""
                return .command(.stopRing(requestId: requestId, target: target))
            case "MESSAGE":
                let target = cmd["target"] as? String ?? ""
                let text = cmd["message"] as? String ?? ""
                return .command(.message(requestId: requestId, target: target, text: text))
            case "PING":
                return .command(.ping(requestId: requestId))
            default:
                return nil
            }

        case "device_list": return .deviceList(jsonData)
        case "status_update": return .statusUpdate(jsonData)
        case "error":
            return .error(obj["message"] as? String ?? "unknown_error")
        default:
            return nil
        }
    }

    static func hello(deviceId: String, token: String) -> String {
        let obj: [String: Any] = [
            "kind": "hello", "deviceId": deviceId, "token": token, "deviceInfo": [:],
        ]
        return jsonString(obj)
    }

    static func ack(requestId: String, status: CommandStatus, detail: String? = nil) -> String {
        var obj: [String: Any] = ["kind": "ack", "requestId": requestId, "status": status.rawValue]
        if let detail { obj["detail"] = detail }
        return jsonString(obj)
    }

    static func heartbeat(batteryPercent: Int?, charging: Bool?) -> String {
        var obj: [String: Any] = ["kind": "heartbeat"]
        if let batteryPercent { obj["battery"] = batteryPercent }
        if let charging { obj["charging"] = charging }
        return jsonString(obj)
    }

    private static func jsonString(_ obj: [String: Any]) -> String {
        let data = (try? JSONSerialization.data(withJSONObject: obj)) ?? Data()
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}
